scripts/popup.js

```diff
@@ -1,217 +1,323 @@
+  function getInputType(apiType) {
+    switch (apiType) {
+      case 'string':
+        return 'text';
+      case 'date':
+        return 'date';
+      default:
+        return null;
     }
-    blob = new Blob([(new Uint8Array([0xEF, 0xBB, 0xBF])), request.content], {
-      type: 'text/html',
-    });
+  }
 
-    postDataToApi(blob);
+  // Generates html for metadata field's label
+  function generateMetadataLabel(metadataField) {
+    return $(`<label>${metadataField.name}</label>`, {
+      htmlFor: metadataField.id,
+    });
   }
-  if (request.processError) {
-    // singlefile.ui.notifyProcessError(request.tabId);
+
+  // Generates html for metadata field's input field
+  function generateMetadataInput(metadataField) {
+    return $('<input />',
+      {
+        id: metadataField.id.split(' ').join('_'), // ID currently has spaces in it, which is no good for html ids
+        name: metadataField.name,
+        type: getInputType(metadataField.type),
+      });
   }
-});
 
-function getInputType(apiType) {
-  switch (apiType) {
-    case 'string':
-      return 'text';
-    case 'date':
-      return 'date';
-    default:
-      return null;
+  // Generates html for metadata field and returns as array
+  function generateMetadataInputs(metadataFields) {
+    const metadataInputs = [];
+    metadataFields.forEach((metadataField) => {
+      metadataField.id = metadataField.name.toLowerCase();
+
+      const container = $('<div class="metadata_item"></div>');
+      container.prepend(generateMetadataInput(metadataField));
+      container.prepend(generateMetadataLabel(metadataField));
+
+      metadataInputs.push(container);
+    });
+
+    return metadataInputs;
   }
-}
 
-function generateMetadataLabel(metadataField) {
-  return $(`<label>${metadataField.name}</label>`, {
-    htmlFor: metadataField.id,
-  });
-}
-
-function generateMetadataInput(metadataField) {
-  return $('<input />',
-    {
-      id: metadataField.id,
-      name: metadataField.name,
-      type: getInputType(metadataField.type),
+  // Adds the given elements to the beginning of the form
+  function prependElementsToForm(elements) {
+    const form = $('#metadata-form');
+
+    elements.forEach((element) => {
+      form.prepend(element);
     });
-}
+  }
 
-function generateMetadataInputs(metadataJson) {
-  const metadataInputs = [];
+  // Adds the given element to the end of the form
+  function prependElementToForm(element) {
+    const form = $('#metadata-form');
 
-  metadataJson.forEach((metadataField) => {
-    metadataField.id = metadataField.name.toLowerCase();
+    form.prepend(element);
+  }
 
-    const inputHtml = generateMetadataInput(metadataField);
-    const labelHtml = generateMetadataLabel(metadataField);
+  function generateFormHtml(groupData) {
+    const sections = [];
+    const checkboxDiv = $('<div class="metadata-checkboxes"><h1>Metadata Groups</h1></div>');
+    const defaultGroups = ['Generic'];
```

Comment:
I was planning on having it in there, just left it out for testing purposes for the checkboxes showing and hiding.